src/dfx/src/config/cache.rs

```diff
@@ -286,57 +220,42 @@ pub fn get_binary_path_from_version(version: &str, binary_name: &str) -> DfxResu
     Ok(get_bin_cache(version)?.join(binary_name))
 }
 
-#[context("Failed to get binary '{}' for version '{}'.", name, version)]
-pub fn binary_command_from_version(version: &str, name: &str) -> DfxResult<std::process::Command> {
+pub fn binary_command_from_version(
+    version: &str,
+    name: &str,
+) -> Result<std::process::Command, CacheError> {
     let path = get_binary_path_from_version(version, name)?;
     let cmd = std::process::Command::new(path);
 
     Ok(cmd)
 }
 
-#[context("Failed to list cache versions.")]
-pub fn list_versions() -> DfxResult<Vec<Version>> {
+pub fn list_versions() -> Result<Vec<Version>, CacheError> {
     let root = get_bin_cache_root()?;
     let mut result: Vec<Version> = Vec::new();
 
-    for entry in std::fs::read_dir(&root).with_context(|| {
-        format!(
-            "Failed to read bin cache root content at {}.",
-            root.to_string_lossy()
-        )
-    })? {
-        let entry = entry.with_context(|| {
-            format!(
-                "Failed to read an entry in bin cache root at {}.",
-                root.to_string_lossy()
-            )
-        })?;
+    for entry in dfx_core::fs::read_dir(&root)? {
+        let entry = entry?;
         if let Some(version) = entry.file_name().to_str() {
             if version.starts_with('_') {
                 // temp directory for version being installed
                 continue;
             }
-            result.push(
-                Version::parse(version)
-                    .with_context(|| format!("Failed to parse version from {}.", version))?,
-            );
+            result.push(Version::parse(version).map_err(CacheError::MalformedSemverVersion)?);
         }
     }
 
     Ok(result)
 }
 
-pub fn call_cached_dfx(v: &Version) -> DfxResult<ExitStatus> {
+pub fn call_cached_dfx(v: &Version) -> Result<ExitStatus, CacheError> {
     let v = format!("{}", v);
     let command_path = get_binary_path_from_version(&v, "dfx")?;
-    if command_path
-        == std::env::current_exe().context("Failed to get currently running executable.")?
-    {
-        bail!("Invalid cache for version {}.", v)
+    if command_path == dfx_core::foundation::get_current_exe()? {
+        return Err(CacheError::UnknownVersion(v));
     }
 
-    std::process::Command::new(command_path)
-        .args(std::env::args().skip(1))
-        .status()
-        .map_err(DfxError::from)
+    let cmd = std::process::Command::new(command_path)
+        .args(std::env::args().skip(1));
+    dfx_core::process::execute_process(cmd).map_err(CacheError::ProcessError)
```

Comment:
Since `ProcessError` is transparent:
```suggestion
    let result = dfx_core::process::execute_process(cmd)?;
    Ok(result)
```